# Assignment 21: 

## Logistic Regression: 1. Bank Loan Modelling

In [1]:
import pandas as pd
import numpy as np

In [11]:
data = pd.read_excel("Bank_Personal_Loan_Modelling.xlsx",sheet_name=1)
data.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


In [9]:
data.describe().T

count          mean          std     min       25%  \
ID                  5000.0   2500.500000  1443.520003     1.0   1250.75   
Age                 5000.0     45.338400    11.463166    23.0     35.00   
Experience          5000.0     20.104600    11.467954    -3.0     10.00   
Income              5000.0     73.774200    46.033729     8.0     39.00   
ZIP Code            5000.0  93152.503000  2121.852197  9307.0  91911.00   
Family              5000.0      2.396400     1.147663     1.0      1.00   
CCAvg               5000.0      1.937913     1.747666     0.0      0.70   
Education           5000.0      1.881000     0.839869     1.0      1.00   
Mortgage            5000.0     56.498800   101.713802     0.0      0.00   
Personal Loan       5000.0      0.096000     0.294621     0.0      0.00   
Securities Account  5000.0      0.104400     0.305809     0.0      0.00   
CD Account          5000.0      0.060400     0.238250     0.0      0.00   
Online              5000.0      0.596800     0.490589     0.0      0.00   
CreditCard          5000.0      0.294000     0.455637     0.0      0.00   

                        50%       75%      max  
ID                   2500.5   3750.25   5000.0  
Age                    45.0     55.00     67.0  
Experience             20.0     30.00     43.0  
Income                 64.0     98.00    224.0  
ZIP Code            93437.0  94608.00  96651.0  
Family                  2.0      3.00      4.0  
CCAvg                   1.5      2.50     10.0  
Education               2.0      3.00      3.0  
Mortgage                0.0    101.00    635.0  
Personal Loan           0.0      0.00      1.0  
Securities Account      0.0      0.00      1.0  
CD Account              0.0      0.00      1.0  
Online                  1.0      1.00      1.0  
CreditCard              0.0      1.00      1.0

In [17]:
data.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [32]:
y=data['Personal Loan'] #defining the dependent variable

In [33]:
x=data[[ 'Age', 'Experience', 'Income', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Securities Account',
       'CD Account', 'Online', 'CreditCard']] # all the independent variables

In [34]:
import statsmodels.api as sm 

In [35]:
x1=sm.add_constant(x) #adding the b0 (beta not) constant value to the equation

In [36]:
logistic = sm.Logit(y,x1)

In [37]:
result=logistic.fit()

Optimization terminated successfully.
         Current function value: 0.128435
         Iterations 9


In [38]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:          Personal Loan   No. Observations:                 5000
Model:                          Logit   Df Residuals:                     4988
Method:                           MLE   Df Model:                           11
Date:                Tue, 05 Jan 2021   Pseudo R-squ.:                  0.5938
Time:                        03:12:53   Log-Likelihood:                -642.18
converged:                       True   LL-Null:                       -1581.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                -12.1928      1.645     -7.411      0.000     -15.417      -8.968
Age                   -0.0536      0.061     -0.874      0.382      -0.174       0.067
Experience             0.0638      0.061      1.046      0.295      -0.056       0.183
Income                 0.0546      0.003     20.831      0.000       0.049       0.060
Family                 0.6958      0.074      9.364      0.000       0.550       0.841
CCAvg                  0.1240      0.040      3.127      0.002       0.046       0.202
Education              1.7362      0.115     15.088      0.000       1.511       1.962
Mortgage               0.0005      0.001      0.856      0.392      -0.001       0.002
Securities Account    -0.9368      0.286     -3.277      0.001      -1.497      -0.377
CD Account             3.8225      0.324     11.800      0.000       3.188       4.457
Online                -0.6752      0.157     -4.298      0.000      -0.983      -0.367
CreditCard            -1.1197      0.205     -5.462      0.000      -1.522      -0.718
======================================================================================
"""

__From the above table we can infer that__:

- those values with __p_value__ as __0__ are most significant in deciding the outcome of the dependent variable like, Const, Income, Family, Education, CD Account, online, Credit Card.


- other values which are __<0.05__ are significantly important & those __>0.05__ are insignificant (Age, Experience,Mortagage)

 

### Probability of getting the loan 

In [96]:
# checking the probability with least significant IDV
# For Age lets check the loan outcome & assume the age =28
import math
prob = 1/(1+(math.exp(12.1928 +0.0536*28)))
prob

1.1296517826778067e-06

__The probability for someone with age 28 is 0.00011299 % which is very less__

In [97]:
# checking the probability with most significant IDV
# For Age lets check the loan outcome & assume the income=155
import math
prob1 = 1/(1+(math.exp(12.1928 -0.0546*150)))
prob2 = 1/(1+(math.exp(12.1928 -0.0546*240)))
prob3 = 1/(1+(math.exp(12.1928 -0.0546*300)))
print(prob1)
print(prob2)
print(prob3)

0.017936821074111993
0.7132456570912225
0.9850384927519747


__We can observe here that, as the income values increases from 150, 240 to 300 the probability of getting a loan also increases accordingly from 1.79% >> 71.3% >> to 98.5% respectively__

## Logistic Regression: 2. Attrition Analysis

In [188]:
data1 = pd.read_csv("general_data.csv")
data1.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [189]:
data1.shape

(4410, 24)

In [190]:
data1.isnull().count()

Age                        4410
Attrition                  4410
BusinessTravel             4410
Department                 4410
DistanceFromHome           4410
Education                  4410
EducationField             4410
EmployeeCount              4410
EmployeeID                 4410
Gender                     4410
JobLevel                   4410
JobRole                    4410
MaritalStatus              4410
MonthlyIncome              4410
NumCompaniesWorked         4410
Over18                     4410
PercentSalaryHike          4410
StandardHours              4410
StockOptionLevel           4410
TotalWorkingYears          4410
TrainingTimesLastYear      4410
YearsAtCompany             4410
YearsSinceLastPromotion    4410
YearsWithCurrManager       4410
dtype: int64

In [191]:
data1.isna().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [192]:
data1['NumCompaniesWorked'] = data1['NumCompaniesWorked'].fillna(0)

In [193]:
data1.isna().sum()

Age                        0
Attrition                  0
BusinessTravel             0
Department                 0
DistanceFromHome           0
Education                  0
EducationField             0
EmployeeCount              0
EmployeeID                 0
Gender                     0
JobLevel                   0
JobRole                    0
MaritalStatus              0
MonthlyIncome              0
NumCompaniesWorked         0
Over18                     0
PercentSalaryHike          0
StandardHours              0
StockOptionLevel           0
TotalWorkingYears          9
TrainingTimesLastYear      0
YearsAtCompany             0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
dtype: int64

In [194]:
# pd.get_dummies(data1)

In [195]:
data1['BusinessTravel'].unique()

array(['Travel_Rarely', 'Travel_Frequently', 'Non-Travel'], dtype=object)

In [196]:
from sklearn import preprocessing
le=preprocessing.LabelEncoder()
data1['BusinessTravel']= le.fit_transform(data1['BusinessTravel'])
data1['BusinessTravel'].unique()

array([2, 1, 0])

In [197]:
## Perform label encoding across all the dataframe

In [198]:
# data1.apply(le.fit_transform) this aplies to all the columns & we need to apply for only categorical 

In [199]:
data1.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [200]:
# applying t selected columns
data1[['Attrition','Department',
       'Education','EducationField', 'Gender', 'JobRole', 'MaritalStatus', 
        'Over18']]=data1[['Attrition','Department',
       'Education','EducationField', 'Gender', 'JobRole', 'MaritalStatus', 
        'Over18']].apply(le.fit_transform)

In [201]:
data1.head()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0   51          0               2           2                 6          1   
1   31          1               1           1                10          0   
2   32          0               1           1                17          3   
3   38          0               0           1                 2          4   
4   32          0               2           1                10          0   

   EducationField  EmployeeCount  EmployeeID  Gender  ...  NumCompaniesWorked  \
0               1              1           1       0  ...                 1.0   
1               1              1           2       0  ...                 0.0   
2               4              1           3       1  ...                 1.0   
3               1              1           4       1  ...                 3.0   
4               3              1           5       1  ...                 4.0   

   Over18  PercentSalaryHike  StandardHours  StockOptionLevel  \
0       0                 11              8                 0   
1       0                 23              8                 1   
2       0                 15              8                 3   
3       0                 11              8                 3   
4       0                 12              8                 2   

   TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                1.0                      6               1   
1                6.0                      3               5   
2                5.0                      2               5   
3               13.0                      5               8   
4                9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [202]:
data1.describe().T

count          mean           std      min  \
Age                      4410.0     36.923810      9.133301     18.0   
Attrition                4410.0      0.161224      0.367780      0.0   
BusinessTravel           4410.0      1.607483      0.665304      0.0   
Department               4410.0      1.260544      0.527673      0.0   
DistanceFromHome         4410.0      9.192517      8.105026      1.0   
Education                4410.0      1.912925      1.023933      0.0   
EducationField           4410.0      2.247619      1.331067      0.0   
EmployeeCount            4410.0      1.000000      0.000000      1.0   
EmployeeID               4410.0   2205.500000   1273.201673      1.0   
Gender                   4410.0      0.600000      0.489954      0.0   
JobLevel                 4410.0      2.063946      1.106689      1.0   
JobRole                  4410.0      4.458503      2.461263      0.0   
MaritalStatus            4410.0      1.097279      0.729955      0.0   
MonthlyIncome            4410.0  65029.312925  47068.888559  10090.0   
NumCompaniesWorked       4410.0      2.683220      2.499737      0.0   
Over18                   4410.0      0.000000      0.000000      0.0   
PercentSalaryHike        4410.0     15.209524      3.659108     11.0   
StandardHours            4410.0      8.000000      0.000000      8.0   
StockOptionLevel         4410.0      0.793878      0.851883      0.0   
TotalWorkingYears        4401.0     11.279936      7.782222      0.0   
TrainingTimesLastYear    4410.0      2.799320      1.288978      0.0   
YearsAtCompany           4410.0      7.008163      6.125135      0.0   
YearsSinceLastPromotion  4410.0      2.187755      3.221699      0.0   
YearsWithCurrManager     4410.0      4.123129      3.567327      0.0   

                              25%      50%       75%       max  
Age                         30.00     36.0     43.00      60.0  
Attrition                    0.00      0.0      0.00       1.0  
BusinessTravel               1.00      2.0      2.00       2.0  
Department                   1.00      1.0      2.00       2.0  
DistanceFromHome             2.00      7.0     14.00      29.0  
Education                    1.00      2.0      3.00       4.0  
EducationField               1.00      2.0      3.00       5.0  
EmployeeCount                1.00      1.0      1.00       1.0  
EmployeeID                1103.25   2205.5   3307.75    4410.0  
Gender                       0.00      1.0      1.00       1.0  
JobLevel                     1.00      2.0      3.00       5.0  
JobRole                      2.00      5.0      7.00       8.0  
MaritalStatus                1.00      1.0      2.00       2.0  
MonthlyIncome            29110.00  49190.0  83800.00  199990.0  
NumCompaniesWorked           1.00      2.0      4.00       9.0  
Over18                       0.00      0.0      0.00       0.0  
PercentSalaryHike           12.00     14.0     18.00      25.0  
StandardHours                8.00      8.0      8.00       8.0  
StockOptionLevel             0.00      1.0      1.00       3.0  
TotalWorkingYears            6.00     10.0     15.00      40.0  
TrainingTimesLastYear        2.00      3.0      3.00       6.0  
YearsAtCompany               3.00      5.0      9.00      40.0  
YearsSinceLastPromotion      0.00      1.0      3.00      15.0  
YearsWithCurrManager         2.00      3.0      7.00      17.0

In [203]:
data1=pd.get_dummies(data1)
data1.head()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0   51          0               2           2                 6          1   
1   31          1               1           1                10          0   
2   32          0               1           1                17          3   
3   38          0               0           1                 2          4   
4   32          0               2           1                10          0   

   EducationField  EmployeeCount  EmployeeID  Gender  ...  NumCompaniesWorked  \
0               1              1           1       0  ...                 1.0   
1               1              1           2       0  ...                 0.0   
2               4              1           3       1  ...                 1.0   
3               1              1           4       1  ...                 3.0   
4               3              1           5       1  ...                 4.0   

   Over18  PercentSalaryHike  StandardHours  StockOptionLevel  \
0       0                 11              8                 0   
1       0                 23              8                 1   
2       0                 15              8                 3   
3       0                 11              8                 3   
4       0                 12              8                 2   

   TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                1.0                      6               1   
1                6.0                      3               5   
2                5.0                      2               5   
3               13.0                      5               8   
4                9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [204]:
y2=data1['Attrition']

In [205]:
x2=data1[['Age', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager']]

In [212]:
# x3=sm.add_constant(x2) 

In [213]:
# logistic1 = sm.Logit(y2,x3)